stdlib/type.mli

```diff
@@ -35,3 +35,86 @@ type (_, _) eq = Equal: ('a, 'a) eq
 
     At runtime, this function simply returns its second argument unchanged.
 *)
+
+(** {1:identifiers Type identifiers} *)
+
+(** Type identifiers.
+
+    A type identifier is a value that denotes a type. Given two type
+    identifiers they can be tested for {{!Id.equal}equality} to prove
+    they denote the same type. Note that:
+
+    - Unequal identifiers does not imply unequal types: a given type can be
+      denoted by more than one identifier.
+    - Type identifiers can be marshalled, but they get a new, distinct,
+      identity on unmarshalling, so the equalities are lost.
+
+    See an {{!Id.example}example} of use. *)
+module Id : sig
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
+  (** The type for identifiers for type ['a]. *)
+
+  val make : unit -> 'a t
+  (** [make ()] is a new type identifier. *)
+
+  val equal : 'a t -> 'b t -> ('a, 'b) eq option
+  (** [equal id0 id1] is [Some Equal] if identifier [id0] is equal to [id1]
+      and [None] otherwise. *)
+
+  val uid : 'a t -> int
+  (** [uid id] is a runtime unique identifier for [id]. *)
+
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to
+      implement a simple heterogeneous key-value dictionary which you
+      can see as an extensible record type. In contrast to
```

Comment:
Nitpicking, but personally I would remove `which you can see as an extensible record type`; I feel it makes the explanation a bit more complicated than it could be...
```suggestion
      implement a simple heterogeneous key-value dictionary. In contrast to
```